File: benchmarks/70-cobyqa-constrained/cobyqa_constrained.py
#!/usr/bin/env python3
"""Experiment 70 -- COBYQA (derivative-free quadratic, constrained) (scipy column).

Same 4 problems as ``cobyqa_constrained.m``, same labels, same order, solved with
scipy.optimize.minimize(method="COBYQA") (Ragonneau & Zhang 2023). Both solvers
are derivative-free (no Jacobian either side). Unlike COBYLA, COBYQA handles
equality constraints natively, so this suite mixes eq + ineq + bounds. The check
is Round[10^3 * f] at the optimum (10^2 for HS71, whose optimum is irrational);
both must agree or the join CHECK-FAILs.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
from scipy.optimize import minimize

from harness import bench, check, require

require(["scipy.optimize:minimize"])

_OPTS = {"maxiter": 3000}


# ---- 01 HS71: min x1 x4 (x1+x2+x3) + x3; ineq x1x2x3x4>=25; eq sumsq==40; 1<=xi<=5
def _hs71(v):
    x1, x2, x3, x4 = v
    return x1 * x4 * (x1 + x2 + x3) + x3
_hs71_cons = [
    {"type": "ineq", "fun": lambda v: v[0] * v[1] * v[2] * v[3] - 25.0},
    {"type": "eq",   "fun": lambda v: v[0] ** 2 + v[1] ** 2 + v[2] ** 2 + v[3] ** 2 - 40.0},
]

def _solve_hs71():
    return minimize(_hs71, [1.0, 5.0, 5.0, 1.0], method="COBYQA",
                    bounds=[(1.0, 5.0)] * 4, constraints=_hs71_cons, options=_OPTS)

bench("01 HS71", _solve_hs71)
check("01 HS71", int(np.floor(1e2 * _solve_hs71().fun + 0.5)))


# ---- 02 nonlinear equality: min x^2+y^2 s.t. x y == 1 -> 2 at (1,1)
def _nl(v):   return v[0] ** 2 + v[1] ** 2
_nl_cons = [{"type": "eq", "fun": lambda v: v[0] * v[1] - 1.0}]

def _solve_nl():
    return minimize(_nl, [1.5, 0.8], method="COBYQA", constraints=_nl_cons, options=_OPTS)

bench("02 Nonlinear-eq xy=1", _solve_nl)
check("02 Nonlinear-eq xy=1", int(np.floor(1e3 * _solve_nl().fun + 0.5)))


# ---- 03 constrained Rosenbrock: min (1-x)^2+100(y-x^2)^2 s.t. x+y<=1
def _cros(v):  x, y = v; return (1 - x) ** 2 + 100 * (y - x * x) ** 2
_cros_cons = [{"type": "ineq", "fun": lambda v: 1.0 - v[0] - v[1]}]

def _solve_cros():
    return minimize(_cros, [-1.0, 1.0], method="COBYQA", constraints=_cros_cons, options=_OPTS)

bench("03 Constrained-Rosenbrock", _solve_cros)
check("03 Constrained-Rosenbrock", int(np.floor(1e3 * _solve_cros().fun + 0.5)))


# ---- 04 equal-weight simplex QP n=8: min sum w^2 s.t. sum w == 1, w >= 0 -> 1/8
def _make_simplex(n):
    fun = lambda v: float(np.dot(v, v))
    cons = [{"type": "eq", "fun": lambda v: float(np.sum(v)) - 1.0}]
    x0 = np.full(n, 1.0 / n)
    bounds = [(0.0, None)] * n

    def solve():
        return minimize(fun, x0, method="COBYQA", bounds=bounds, constraints=cons, options=_OPTS)
    return solve

_s8 = _make_simplex(8)
bench("04 Simplex QP n8", _s8)
check("04 Simplex QP n8", int(np.floor(1e3 * _s8().fun + 0.5)))

File: benchmarks/70-cobyqa-constrained/cobyqa_constrained.m
(* Experiment 70 -- COBYQA (derivative-free quadratic-approximation, constrained)
   optimization (FindMinimum, Method -> "COBYQA").

   MEASURES src/numerical_calculus/findmin.c's derivative-free trust-region SQP
   with QUADRATIC interpolation models -- full quadratic models of f and every
   constraint built by finite differences on a structured stencil, a
   Byrd-Omojokun-style trust-region SQP step (the same active-set QP as SLSQP),
   an L1 penalty merit, and native equality + inequality + bound handling --
   against scipy.optimize.minimize(method="COBYQA") (Ragonneau & Zhang 2023).

   Both solvers are DERIVATIVE-FREE (no Jacobian either side). Unlike COBYLA
   (experiment 69, inequality-only), COBYQA handles equalities natively, so this
   suite mixes eq + ineq + bounds: HS71, a nonlinear equality, the equal-weight
   simplex QP, and constrained Rosenbrock (the curved valley COBYQA's quadratic
   models navigate where COBYLA's linear ones stall). The check is Round[10^3 * f]
   at the optimum (10^2 for HS71, whose optimum is irrational): both must reach it
   or the join CHECK-FAILs. Objectives/constraints are built with Total[...] over
   fresh global symbols and passed through Evaluate[] past FindMinimum's HoldAll. *)

Get["../harness.m"];

require[{"FindMinimum"}];

(* Equal-weight simplex QP: min Total[w^2] s.t. Total[w]==1, w_i>=0 -> 1/n. *)
splxVars[n_] := Table[Symbol["w" <> ToString[i]], {i, 1, n}];
splxProb[n_] := With[{w = splxVars[n]}, {Total[w^2], Total[w] == 1 && And @@ (# >= 0 & /@ w)}];
splxSpec[n_] := Table[{Symbol["w" <> ToString[i]], 1./n}, {i, 1, n}];
splx8 = splxProb[8]; splx8s = splxSpec[8];

(* ---- 01 HS71: eq + ineq + bounds, f* = 17.0140 at (1, 4.743, 3.821, 1.379) ---- *)
hs71 = {x1 x4 (x1 + x2 + x3) + x3,
        x1 x2 x3 x4 >= 25 && x1^2 + x2^2 + x3^2 + x4^2 == 40 &&
        1 <= x1 <= 5 && 1 <= x2 <= 5 && 1 <= x3 <= 5 && 1 <= x4 <= 5};
bench["01 HS71", FindMinimum[Evaluate[hs71], {{x1, 1}, {x2, 5}, {x3, 5}, {x4, 1}}, Method -> "COBYQA", MaxIterations -> 3000];];
check["01 HS71", Round[10^2 First[FindMinimum[Evaluate[hs71], {{x1, 1}, {x2, 5}, {x3, 5}, {x4, 1}}, Method -> "COBYQA", MaxIterations -> 3000]]]];

(* ---- 02 Nonlinear equality: min x^2+y^2 s.t. x y == 1 -> f* = 2 at (1,1).
        Start (1.5, 0.8): scipy COBYQA is start-sensitive on this hyperbola and
        stalls at the (2, 0.5) start; (1.5, 0.8) reaches the optimum in both. ---- *)
nlEq = {x^2 + y^2, x y == 1};
bench["02 Nonlinear-eq xy=1", FindMinimum[Evaluate[nlEq], {{x, 1.5}, {y, 0.8}}, Method -> "COBYQA", MaxIterations -> 3000];];
check["02 Nonlinear-eq xy=1", Round[10^3 First[FindMinimum[Evaluate[nlEq], {{x, 1.5}, {y, 0.8}}, Method -> "COBYQA", MaxIterations -> 3000]]]];

(* ---- 03 Constrained Rosenbrock: min (1-x)^2+100(y-x^2)^2 s.t. x+y<=1.
        On the line x+y==1 (unconstrained min (1,1) has x+y=2>1). The curved
        valley COBYQA's quadratic models navigate; COBYLA's linear ones stall. ---- *)
cros = {(1 - x)^2 + 100 (y - x^2)^2, x + y <= 1};
bench["03 Constrained-Rosenbrock", FindMinimum[Evaluate[cros], {{x, -1}, {y, 1}}, Method -> "COBYQA", MaxIterations -> 3000];];
check["03 Constrained-Rosenbrock", Round[10^3 First[FindMinimum[Evaluate[cros], {{x, -1}, {y, 1}}, Method -> "COBYQA", MaxIterations -> 3000]]]];

(* ---- 04 Simplex QP n=8, f* = 1/8 = 0.125 ---- *)
bench["04 Simplex QP n8", FindMinimum[Evaluate[splx8], Evaluate[splx8s], Method -> "COBYQA", MaxIterations -> 3000];];
check["04 Simplex QP n8", Round[10^3 First[FindMinimum[Evaluate[splx8], Evaluate[splx8s], Method -> "COBYQA", MaxIterations -> 3000]]]];
